### ISSUE TITLE
Handle `DirichletBC` dependencies in `EquationSolver`

### ISSUE THREAD

**Issue by jrmaddison** [tlm-adjoint/tlm_adjoint#422]
Some options:
- Allow a `DirichletBC` to be a variable.
- Handle only the homogeneous case in `EquationSolver`, and lift boundary conditions when annotating operations.

Some difficulties:

- Order of boundary conditions matters, if they set the same degrees of freedom.
- In principle the `function_arg` property could appear _both_ in the `Form` _and_ be used to set boundary values.
- In principle the `function_arg` property could appear _both_ as the solution _and_ be used to set boundary values. This should probably raise an exception.

**Comment by jrmaddison** [comment]
I've investigated treating `DirichletBC.function_arg` properties as dependencies:

- It is possible, but looks like it's equivalent to lifting (at least in the linear variational problem case), with the adjoint terms generated by lifting instead computed within `EquationSolver`. 
- Some of the complexity can be handled with careful modification of the residual, but overall the implementation looks like it would still be quite complicated, particularly due to the boundary condition ordering problem.
- I managed to get a basic test working at first order, but not at second order, possibly due to incorrect derivatives of the modified residual.

It looks like it is much simpler to lift in the patches, and add a warning in the `EquationSolver` constructor to note the missing computational graph edge if a `DirichletBC.function_arg` property is a `Function`.
